std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,228 @@ unittest
                 2, "two",
                 ));
 }
+
+/**
+Returns true if all of the elements in $(D r1) are present in $(D r2). Otherwise,
+returns false.
+
+If your parameters are forward ranges, then this function can allocate if the
+$(D AllocateGC.yes) flag is passed. This has the benefit of being much
+faster than the $(D AllocateGC.no) option.
+
+Or, if your two ranges are sortable, then passing sorted ranges will be faster
+than the other two options. If your sorted ranges need custom equality, you
+can override the default $(D "a == b") predicate. See $(D binaryFun) for more
+details
+
+Non allocating forward range option: $(BIGOH r1.length * r2.length)
+Allocating forward range option: $(BIGOH r1.length) + $(BIGOH r2.length)
+Non allocating sorted range option: $(BIGOH min(r1.length, r2.length))
+
+Params:
+    r1 = Any finite forward range
+    r2 = Any finite forward range
+
+Returns:
+    $(D bool)
+*/
+
+alias AllocateGC = Flag!"AllocateGC";
+
+bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2, AllocateGC allocate_gc = AllocateGC.no)
+    if (isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    // If the two ranges aren't the same length, then they cannot
+    // possibly be permutations
+    static if (hasLength!(Range1) && hasLength!(Range2))
+    {
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
+    }
+
+    // Skip the elements at the beginning where r1.front == r2.front,
+    // they are in the same order and don't need to be counted.
+    while (!r1.empty && !r2.empty)
+    {
+        if (r1.front == r2.front)
+        {
+            r1.popFront;
+            r2.popFront;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    if (r1.empty && r2.empty)
+    {
+        return true;
+    }
+
+    alias E1 = Unqual!(ElementType!Range1);
+    alias E2 = Unqual!(ElementType!Range2);
+
+    if (allocate_gc == AllocateGC.yes)
+    {
+        int[CommonType!(E1, E2)] counts;
+
+        foreach (item; r1)
+        {
+            ++counts[item];
+        }
+
+        foreach (item; r2)
+        {
+            --counts[item];
+
+            if (counts[item] == 0)
+            {
+                counts.remove(item);
+            }
+        }
+
+        return counts.length == 0;
+    }
+    else
+    {
+        import std.algorithm.searching : count;
+
+        ulong r1_count;
+        ulong r2_count;
+
+        if (r1.save.walkLength != r2.save.walkLength)
```

Comment:
A couple notes here:

1. You are checking length above for `.length` defining ranges. So no need to check in that case.
2. I realized you only need to check the minimum length here. So for instance, if r1.length i 5, and r2.length is 5000, walkLength will go through all 5000 elements, where it only really needs to get to 6 to realize it's not equal. I don't know if there's an easy way to do this without looping in parallel.